lib/queue.js

```diff
@@ -4,33 +4,65 @@ const spq = require('shuffled-priority-queue')
 const { EventEmitter } = require('events')
 const peerInfo = require('./peer-info')
 const timer = require('./bulk-timer')
-
+const noop = () => {}
 module.exports = (opts) => new PeerQueue(opts)
 
+const BACKOFF_S = 1000
+const BACKOFF_M = 5000
+const BACKOFF_L = 15000
+const FORGET_UNRESPONSIVE = 7500
+const FORGET_BANNED = Infinity
+
 class PeerQueue extends EventEmitter {
   constructor (opts = {}) {
     super()
 
     const {
-      requeue = [ 1000, 5000, 15000 ]
+      requeue = [ BACKOFF_S, BACKOFF_M, BACKOFF_L ],
+      forget = {
+        unresponsive: FORGET_UNRESPONSIVE,
+        banned: FORGET_BANNED
+      }
     } = opts
 
+    const [
+      s = BACKOFF_S,
+      m = BACKOFF_M,
+      l = BACKOFF_L
+    ] = requeue
+    const backoff = [s, m, l, ...requeue.slice(3)]
+
+    const {
+      unresponsive = FORGET_UNRESPONSIVE,
+      banned = FORGET_BANNED
+    } = forget
+
     this.destroyed = false
     this._infos = new Map()
     this._queue = spq()
 
     const push = this._push.bind(this)
-
-    this._requeue = requeue.map(ms => timer(ms, push))
+    const release = this._release.bind(this)
+    this._requeue = backoff.map(ms => timer(ms, push))
+    this._remove = [unresponsive, banned].map((ms) => {
+      return ms < Infinity
+        ? timer(ms, release)
+        : { push: noop, destroy: noop } // noop for infinite time
```

Comment:
do we run into any polymorphic trouble with this?

Reply:
with regards to performance? technically maybe because it doesn't contain the full shape of the `timer`  but given the use case I don't think so.. I mean we could run into bugs if there's more interaction built on top of `timer`. We could create a NoopTimer class and inherit timer from it... but I don't see it a big issue at the moment, wdyt?

Reply:
yea this is fine. was mostly curious :)

Reply:
ahahah 👍 - performance testing all of this will be fun because there's a lot about it that's non-linear